File: 백준/Silver/2121. 넷이 놀기/넷이 놀기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
  static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
  static StringTokenizer st;
  private static int N, A, B;
  private static int[][] arr;
  private static int answer;

  public static void main(String[] args) throws NumberFormatException, IOException {
    N = Integer.parseInt(br.readLine());
    st = new StringTokenizer(br.readLine());
    A = Integer.parseInt(st.nextToken());
    B = Integer.parseInt(st.nextToken());

    arr = new int[N][2];
    for (int i = 0; i < N; i++) {
      st = new StringTokenizer(br.readLine());
      arr[i][0] = Integer.parseInt(st.nextToken());
      arr[i][1] = Integer.parseInt(st.nextToken());
    } // 입력완료

    // 정렬하기
    Arrays.sort(arr, (o1, o2) -> {
      if (o1[0] == o2[0]) {
        return o1[1] - o2[1];
      } else {
        return o1[0] - o2[0];
      }
    });

    // 모든 점 탐색 (1번 점)
    for (int i = 0; i < N; i++) {
      // 2번점이 있다면
      if (binarySearch(arr[i][0] + A, arr[i][1]))
        // 3번점이 있다면
        if (binarySearch(arr[i][0], arr[i][1] + B))
          // 4번 점이 있다면
          if (binarySearch(arr[i][0] + A, arr[i][1] + B))
            // 직사각형이 완성
            answer++;
    }
    
    System.out.println(answer);

  }

  private static boolean binarySearch(int targetX, int targetY) {
    int start = 0;
    int end = N - 1;
    int mid = 0;
    while (start <= end) {
      mid = (start + end) / 2;
      if (targetX == arr[mid][0]) { // 목표 x좌표와 같다면
        if (targetY == arr[mid][1]) { // 목표 y좌표와 같은지 보기
          return true;
        } else if (targetY > arr[mid][1]) { 
          start = mid + 1;
        } else if (targetY < arr[mid][1]) {
          end = mid - 1;
        }
      } else if (targetX > arr[mid][0]) {
        start = mid + 1;
      } else if(targetX < arr[mid][0]) {
        end = mid - 1;
      }
    }
    return false;
  } // end of main
} // end of class

File: 백준/Silver/2121. 넷이 놀기/넷이 놀기.py
"""
어떤 점의 좌표가 (x, y)일떄

점들의 리스트 중에

(x + A, y),
(x, y + B),
(x + A, y + B)
인 점이 있는지 확인하기

다만 하나의 점당 3번이나 확인해야하므로 오래걸린다 
따라서 이분탐색으로 시간을 줄이기
"""

import sys

input = lambda: sys.stdin.readline().strip()

N = int(input())

A, B = map(int, input().split())

lst = []
for _ in range(N):
    lst.append(list(map(int, input().split())))

lst = sorted(lst)


def binarySearch(targetX, targetY):
    isFind = False
    start = 0
    end = N-1
    while start <= end:
        mid = (start + end) // 2
        if lst[mid][0] > targetX:
            end = mid - 1
        elif lst[mid][0] < targetX:
            start = mid + 1
        else:
            if lst[mid][1] > targetY:
                end = mid - 1
            elif lst[mid][1] < targetY:
                start = mid + 1
            else:
                isFind = True
                break
    return isFind


answer = 0
for x, y in lst:
    if binarySearch(x + A, y) and binarySearch(x, y + B) and binarySearch(x + A, y + B):
        answer += 1

print(answer)